ark/phenotyping/cluster_helpers.py

```diff
@@ -0,0 +1,261 @@
+                Mh = self.cluster_(n_clusters=k).fit_predict(resample_data)
+                # find indexes of elements from same clusters with bisection
+                # on sorted array => this is more efficient than brute force search
+                index_mapping = np.array((Mh, resampled_indices)).T
+                index_mapping = index_mapping[index_mapping[:, 0].argsort()]
+                sorted_ = index_mapping[:, 0]
+                id_clusts = index_mapping[:, 1]
+                for i in range(k):  # for each cluster
+                    ia = bisect.bisect_left(sorted_, i)
+                    ib = bisect.bisect_right(sorted_, i)
+                    is_ = id_clusts[ia:ib]
+                    ids_ = np.array(list(combinations(is_, 2))).T
+                    # sometimes only one element is in a cluster (no combinations)
+                    if ids_.size != 0:
+                        Mk[i_, ids_[0], ids_[1]] += 1
+                # increment counts
+                ids_2 = np.array(list(combinations(resampled_indices, 2))).T
+                Is[ids_2[0], ids_2[1]] += 1
+            Mk[i_] /= Is+1e-8  # consensus matrix
+            # Mk[i_] is upper triangular (with zeros on diagonal), we now make it symmetric
+            Mk[i_] += Mk[i_].T
+            Mk[i_, range(data.shape[0]), range(
+                data.shape[0])] = 1  # always with self
+            Is.fill(0)  # reset counter
+        self.Mk = Mk
+        # fits areas under the CDFs
+        self.Ak = np.zeros(self.K_-self.L_)
+        for i, m in enumerate(Mk):
+            hist, bins = np.histogram(m.ravel(), density=True)
+            self.Ak[i] = np.sum(h*(b-a) for b, a, h in zip(bins[1:], bins[:-1], np.cumsum(hist)))
+        # fits differences between areas under CDFs
+        self.deltaK = np.array([(Ab-Aa)/Aa if i > 2 else Aa
+                                for Ab, Aa, i in zip(self.Ak[1:], self.Ak[:-1],
+                                                     range(self.L_, self.K_-1))])
+        self.bestK = np.argmax(self.deltaK) + \
+            self.L_ if self.deltaK.size > 0 else self.L_
+
+    def predict(self):
+        """Predicts on the consensus matrix, for best found cluster number.
+
+        Returns:
+            numpy.ndarray:
+                The consensus matrix prediction for `self.bestK`.
+        """
+        assert self.Mk is not None, "First run fit"
+        return self.cluster_(n_clusters=self.bestK).fit_predict(
+            1-self.Mk[self.bestK-self.L_])
+
+    def predict_data(self, data: pd.DataFrame):
+        """Predicts on the data, for best found cluster number
+
+        Args:
+            data (pandas.DataFrame):
+                `(examples,attributes)` format
+
+        Returns:
+            pandas.DataFrame:
+                The data matrix prediction for `self.bestK`.
+        """
+        assert self.Mk is not None, "First run fit"
+        return self.cluster_(n_clusters=self.bestK).fit_predict(
+            data)
+
+
```

Comment:
Just to clarify, is everything above this copied 100% from ZigaSajovic, or did you change any of it?

Reply:
Yup.